test/framework/controlplane_helpers.go

```diff
@@ -154,17 +155,22 @@ type WaitForControlPlaneToBeReadyInput struct {
 // WaitForControlPlaneToBeReady will wait for a control plane to be ready.
 func WaitForControlPlaneToBeReady(ctx context.Context, input WaitForControlPlaneToBeReadyInput, intervals ...interface{}) {
 	By("Waiting for the control plane to be ready")
-	Eventually(func() (bool, error) {
-		controlplane := &controlplanev1.KubeadmControlPlane{}
+	controlplane := &controlplanev1.KubeadmControlPlane{}
+	Eventually(func() (controlplanev1.KubeadmControlPlane, error) {
 		key := client.ObjectKey{
 			Namespace: input.ControlPlane.GetNamespace(),
 			Name:      input.ControlPlane.GetName(),
 		}
 		if err := input.Getter.Get(ctx, key, controlplane); err != nil {
-			return false, err
+			log.Logf("Failed to list the KCP: %+v", err)
```

Comment:
I assume we intentionally print the err with field names? (just asking because usually we just use %v)

I assume we have to log the error here at all because the `Should` doesn't detect the error anymore, right?

There is probably no way to valid both return parameters (i.e. also err != nil )?

Reply:
> I assume we intentionally print the err with field names? (just asking because usually we just use %v)
Good point. I saws other log statements in the file (for ex, (here](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/framework/controlplane_helpers.go#L98), [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/test/framework/controlplane_helpers.go#L52)) uses`+%v` and tried to follow the convention. Should we change other log statements with `%v`?

> I assume we have to log the error here at all because the `Should` doesn't check the error, right?
Yes. `Should` makes an assertion against the first argument. 

> There is probably no way to validate both return parameters (i.e. also err != nil )?
I don't think so.

Reply:
Sounds good. I'm not sure about `%+v`. I wasn't aware we're using it in other cases too :). I'm fine with leaving it as is.

@fabriziopandini opinions?

Reply:
`Eventaully` can actually handle functions that return multiple values. It runs the assertion on the first returned value and checks that all other returned values are zero-valued. In this case it is inherently checking for err to be nil and the expectation will fail otherwise. 
From [gomega docs](https://onsi.github.io/gomega/#:~:text=If%20multple%20values%20are%20returned%20by%20the%20function%2C%20Eventually%20will%20pass%20the%20first%20value%20to%20the%20matcher%20and%20require%20that%20all%20others%20are%20zero-valued.%20This%20allows%20you%20to%20pass%20Eventually%20a%20function%20that%20returns%20a%20value%20and%20an%20error%20-%20a%20common%20pattern%20in%20Go.):
>If multple values are returned by the function, Eventually will pass the first value to the matcher and require that all others are zero-valued. This allows you to pass Eventually a function that returns a value and an error - a common pattern in Go.

Reply:
Sounds nice. The question for us is probably would it validate kcp and the error and print both results, or would it only print KCP because it returns after KCP validation fails.

Reply:
I dont think we would want to log the error as this assertion is asynchronous and we are okay with it failing to get the KCP object a few times. Although it would be nice to wrap the error with a message so that if the eventual assertion fails (error is not nil) we get the message in the logs. (I believe it will print the final error if it is not nil)

Reply:
But current state of the PR is fine for me, don't want to block it because of that. But we might want to open an issue to possibly improve our Eventually's

Reply:
Thinking about it, I agree with @ykakarap that we don't need to log the error here because it is `Eventually`.

> I believe it will print the final error if it is not nil
That will be the case.

Should we also remove the error log messages from other Eventually functions in the file? It doesn't need to be part of this PR but just wondering.

Reply:
If we're sure the error is logged when the eventually fails with a incorrect KCP && err != nil => I think yes

Reply:
>If we're sure the error is logged when the eventually fails with a incorrect KCP && err != nil => I think yes

Yes, only if the final error is logged if it is not nil.

Reply:
I ran a test and here is how t will show up if the error is not nil. 
Timed out after 900.000s.
```
  KCP
  {
    "metadata": {
      "creationTimestamp": null
    },
    "spec": {
      "version": "",
      "machineTemplate": {
        "metadata": {},
        "infrastructureRef": {}
      },
      "kubeadmConfigSpec": {}
    },
    "status": {
      "replicas": 0,
      "updatedReplicas": 0,
      "readyReplicas": 0,
      "unavailableReplicas": 0,
      "initialized": false,
      "ready": false
    }
  }
  
  Error: Unexpected non-nil/non-zero extra argument at index 1:
        <*errors.fundamental>: this is a sample error message that should show up
```

Reply:
Thanks @ykakarap for testing this. I've wrapped the error and the output will look like below.
```
  Error: Unexpected non-nil/non-zero extra argument at index 1:
  	<*errors.errorString>: &errors.errorString{s:"failed to get KCP: this is a sample error message that should show up"}
```